File: de.woq.osgi.java.itestsupport/src/main/java/de/woq/osgi/java/itestsupport/CamelTestSupport.java
package de.woq.osgi.java.itestsupport;

import de.woq.osgi.java.testsupport.XMLMessageFactory;
import org.apache.camel.CamelContext;
import org.apache.camel.Exchange;
import org.apache.camel.ExchangePattern;
import org.apache.camel.Message;
import org.apache.camel.ProducerTemplate;
import org.apache.camel.builder.RouteBuilder;
import org.apache.camel.component.mock.MockComponent;
import org.apache.camel.component.mock.MockEndpoint;
import org.apache.camel.impl.DefaultCamelContext;
import org.apache.camel.impl.DefaultExchange;
import org.apache.camel.impl.DefaultMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

public class CamelTestSupport {

  private CamelContext context = null;

  private final static Logger LOGGER = LoggerFactory.getLogger(CamelTestSupport.class);


  public synchronized void init() throws Exception {

    if (context == null) {
      context = new DefaultCamelContext();

      context.addComponent("mock", new MockComponent());
    }
  }

  public void sendTestMessage(final String message, final Properties properties, final String uri) throws Exception {
    Message msg = null;

    try {
      LOGGER.info("Trying to create message with Message factory [{}]", message);
      msg = new XMLMessageFactory(message).createMessage();
    } catch (Exception e) {
      LOGGER.info("Using text as msg body: [{}]", message);
      msg = new DefaultMessage();
      msg.setBody(message);

      if (properties != null) {
        for(String key: properties.stringPropertyNames() ) {
          LOGGER.info("Setting property [{}] = [{}]", key, properties.getProperty(key));
          msg.setHeader(key, properties.getProperty(key));
        }
      }
    }

    final Exchange sent = new DefaultExchange(getContext(), ExchangePattern.InOnly);
    sent.setIn(msg);

    final ProducerTemplate producer = getContext().createProducerTemplate();
    final Exchange response = producer.send(uri, sent);

    if (response.getException() != null) {
      LOGGER.info("Message not sent to [{}]", uri);
      throw response.getException();
    } else {
      LOGGER.info("Sent test message to [{}]", uri);
    }
  }

  public void sendTestMessage(final String message, final String properties, final String uri) throws Exception {

    Properties props = new Properties();

    if (properties != null) {
      for(String pair : properties.split(";")) {
        String[] keyValue = pair.split("=");
        if (keyValue.length == 2) {
          props.put(keyValue[0].trim(), keyValue[1].trim());
        }
      }
    }

    sendTestMessage(message, props, uri);
  }

  public void sendTestMessage(final String message, final String uri) throws Exception {
    sendTestMessage(message, "", uri);
  }

  public MockEndpoint wireMock(final String mockName, final String uri) throws Exception {

    final MockEndpoint result = (MockEndpoint)getContext().getEndpoint("mock:" + mockName);

    getContext().addRoutes(new RouteBuilder() {
      @Override
      public void configure() throws Exception {
        from(uri).to(result);
      }
    });

    return result;
  }

  public synchronized CamelContext getContext() throws Exception {
    if (context == null) {
      init();
    }
    return context;
  }

}

File: blended.itestsupport/src/main/scala/blended/itestsupport/camel/CamelTestSupport.scala
package blended.itestsupport.camel

import akka.camel.CamelMessage
import blended.testsupport.XMLMessageFactory
import blended.util.FileReader
import org.apache.camel.impl.DefaultExchange
import org.apache.camel.{CamelContext, Exchange, ExchangePattern, Message}
import org.slf4j.LoggerFactory

import scala.collection.convert.Wrappers.JMapWrapper

trait CamelTestSupport {

  val camelContext : CamelContext
  
  def sendTestMessage(message: String, uri: String, binary: Boolean)(implicit context: CamelContext): Either[Exception, CamelMessage] = {
    sendTestMessage(message, Map.empty, uri, binary)
  }

  def sendTestMessage(message: String, properties: Map[String, String], uri: String, binary: Boolean)(implicit context: CamelContext) : Either[Exception, CamelMessage] = {
    sendTestMessage(message, properties, uri, true, binary)
  }

  def sendTestMessage(message: String, properties: Map[String, String], uri: String, evaluateXML: Boolean, binary: Boolean)(implicit context: CamelContext) : Either[Exception, CamelMessage] = {

    val log = LoggerFactory.getLogger(classOf[CamelTestSupport])
    
    val exchange = camelExchange(createMessage(message, properties, evaluateXML, binary))
    exchange.setPattern(ExchangePattern.InOnly)

    val producer = context.createProducerTemplate()
    val response : Exchange = producer.send(uri, exchange)

    if (response.getException != null) {
      log.warn(s"Message not sent to [$uri]")
      Left(response.getException())
    }
    else {
      log.info(s"Sent test message to [$uri]")
      Right(camelMessage(exchange.getIn))
    }
  }

  def executeRequest(message: String, properties : Map[String, String], uri: String, binary: Boolean)(implicit context: CamelContext) : Either[Exception, CamelMessage] = {
    executeRequest(message, properties, uri, true, binary)
  }

  def executeRequest(message: String, properties : Map[String, String], uri: String, evaluateXML: Boolean, binary: Boolean)(implicit context: CamelContext) : Either[Exception, CamelMessage] = {

    val log = LoggerFactory.getLogger(classOf[CamelTestSupport])

    val exchange = camelExchange(createMessage(message, properties, evaluateXML, binary))
    exchange.setPattern(ExchangePattern.InOut)

    val producer = context.createProducerTemplate()
    val response = producer.send(uri, exchange)

    Option(response.getException()) match {
      case Some(e) =>
        log.warn(s"Executing request on [$uri] failed")
        Left(e)
      case None =>
        Right(camelMessage(response.getOut()))
    }
  }

  def headerExists(msg: CamelMessage, headerName: String) = msg.getHeaders.containsKey(headerName)

  def missingHeaderNames(exchange: CamelMessage, mandatoryHeaders: List[String]) =
    mandatoryHeaders.filter( headerName => !headerExists(exchange, headerName))

  def camelMessage(msg: Message)(implicit context: CamelContext) : CamelMessage =
    CamelMessage(msg.getBody, JMapWrapper(msg.getHeaders).mapValues { _.asInstanceOf[Any] }.toMap)

  private [CamelTestSupport] def createMessage(message: String, properties: Map[String, String], evaluateXML: Boolean, binary: Boolean)(implicit context: CamelContext) : CamelMessage =
    (evaluateXML match {
      case true => createMessageFromXML(message, binary)
      case false => createMessageFromFile(message, properties)
    }) match {
      case None =>
        CamelMessage(message, properties)
      case Some(m) => m
    }

  private[CamelTestSupport] def createMessageFromFile(message: String, props: Map[String, String])(implicit context: CamelContext) : Option[CamelMessage] = {
    try {
      val content: Array[Byte] = FileReader.readFile(message)
      Some(CamelMessage(content, props.mapValues { _.asInstanceOf[Any] } ))
    } catch {
      case e: Exception => None
    }
  }

  private[CamelTestSupport] def createMessageFromXML(message: String, binary: Boolean)(implicit context: CamelContext) : Option[CamelMessage] = {
    try {
      binary match {
        case true  => Some(camelMessage(new XMLMessageFactory(message).createBinaryMessage()))
        case false => Some(camelMessage(new XMLMessageFactory(message).createTextMessage()))
      }
    } catch {
      case e: Exception => None
    }
  }


  private[this] def camelExchange(msg: CamelMessage)(implicit context: CamelContext) : Exchange = {

    val exchange = new DefaultExchange(context)
    
    exchange.getIn.setBody(msg.body)
    msg.headers.foreach { case (k, v) => exchange.getIn.setHeader(k, v) }
    
    exchange
  }
  
}
